### ISSUE TITLE
Create default classes and implement into codebase

### ISSUE THREAD

**Issue by jackcasey067** [uchicago-cs/chiventure#867]
The playerclass component should offer some default classes, for two reasons: 
 1. To allow more robust internal testing of chiventure rpg components. 
 2. To allow game designers to use them (or maybe even to augment them) in their games.

Goals:
- Design some **prefab** (prefabricated) classes, and implement them into the codebase.
- Give classes unique stats.
- Give classes skills and skill trees, showcasing unique skills if possible.

**Comment by jackcasey067** [comment]
The branch for this issue is `playerclass/prefab-classes`.  I added `class_prefabs` as a module with its interface exposed in the `/includes` directory, and I added a test file.  These files contain really rudimentary code for the first class, warrior.  It doesn't even have stats yet, so it is definitely subject to change.

It is important to note that we want to merge PR (#833) as soon as it is approved since we will be dealing with skills in our code.

**Comment by jackcasey067** [comment]
Some notes to self, about the environment our code lives in:
- `player_t` has `class_t *player_class` as a member, and they also have level, health, XP, and inventory.  They access info on stats and effects and skills and so on through `class_t`.  
- We need to have stats for our class, somehow.  Stats must be declared globally, via `stats_global_hash_t`.  We can technically make our own, but that seems like a truly horrid idea.  Instead, we need to access and add to it via `{the chiventure_ctx_t}->game->curr_stats`, which we might even have to initialize.  A similar process is required to access effects, even though we might not do that immediately.
- We require access to the `ctx`object, somehow.

**Comment by jackcasey067** [comment]
Merged `dev` into `playerclass/prefab-classes`, which is significant since dev contains the skill tree integration changes.  We will strive to give these classes skills and skill trees as a result, so we may modify some of the code from that change to do so.

**Comment by jackcasey067** [comment]
I made major organizational changes today.  I am just going to copy the commit message:

> Reorganized class.c, class_prefabs.c, and class_skills.c:
> - class.c:
>   - moved class_init_skills() (or class_skills_init(), I forget the name) to class_prefabs.
>   - added in its place class_add_skills(), which takes a precreated skilltree and skill inventories and adds them to a class.
>   - Restyled the file to my liking.
> - class_skills.c:
>   - The main function there became class_allocate_skills() in class_prefabs.c
>   - The secondary function also moved to class_prefabs_add_skills.
> - class_prefabs:
>   - Received afformentioned functions
> - Also modified .h files, Cmake files, and a test case.
> - WHY:
>   - The functions moved to class_prefab only worked for prefab classes.  Custom classes needs a different way to add its own
>     external trees and skill inventories, which was added.
>   - In fact, the custom class function might end up being used by the prefab one!
> 
Somehow I missed this, but this resulted in the deletion of class_skills.c and .h.

**Comment by jackcasey067** [comment]
Max and I planned out some things in a meeting today.

Right now, we are exposing two main `prefab` functions:
 - 1 that allocates and fills in non-skill-related fields.
 - 1 that fills in the skill-related fields for a pre-allocated prefab class struct.
We debated whether we should have two functions or just one, or have a third that fills in stat-related fields.  However, we will be able to easily switch between these models later on if need be.

Also: 
 - We are interested in allowing users to overwrite certain fields, but this won't necessarily make it into this issue.
 - We are taking on the role of updating wiki articles on our prefab classes. We might also include possibilities to improve them in the future, particularly as skills and skill trees become more complex.

**Comment by jackcasey067** [comment]
Recent work has been going well: 

The framework for defining prefab classes is now in place (this is not yet the case for skills / skill-related fields)
Warrior has been defined, and tests on it succeed.  We simply need to do the other classes like a warrior.

Currently, attributes are undefined, and effects are undefined, and they will probably stay that way.  I know how to do both of these things, but it seems like effects were meant to store current modifiers (which are now being stored by player_t pending an issue in action management), and attributes are hard to meaningfully set when we don't know what kind of world these classes are being dropped into.

**Comment by jackcasey067** [comment]
Similar to above: The framework for skills is done, but initializing skill trees is necessarily a more involved process, so less is refactored into functions.  Still, adding skills to the other classes should be much easier now that the first one is done.

As far as I know, only damage skills are supported by skill trees at the moment, but it's complicated).  I sent them a message asking if this is true.

**Comment by jackcasey067** [comment]
The skill-trees team is working on how skills are being implemented(#855), so it might be prudent to leave some of the skill initialization incomplete until they finish working on this.

**Comment by maxsamengo** [comment]
Updated the Wiki page, _Player Class: Example Stats and Abilities_ to serve as a reference for all prefab classes:
 - Extended the wiki page for all prefab classes. Previously only Sorcerer, Warrior, and Ranger were accounted for. 
 - Added a blurb/rationale for the class stat distribution for each class. Also relevant for the short and long descriptions of the classes. 
 - Renamed "Magic Energy" to "Mana" because it better reflects its role. 

Updated the `class_prefab_new` function for the remaining prefab classes:
- All prefab classes have a base stat total of 110 for the sake of balance. 
- All prefab class stats are multiples of 5, from 5 to 25. This is so that classes are clearly defined and distinguished in terms of stats, and do not include messy stat totals which were previously reflected on the wiki. 
- Removed prefab classes: `Elementalist`, `Knight`, `Druid`, `Sorcerer` because they were too similar to existing prefab classes, and could not be meaningfully unique in terms of stats. These classes's unique themes (Druid -> nature, etc.) can not currently be reflected through the limited number of stats that we support.

**Comment by jackcasey067** [comment]
Closing Statement:

I am closing this issue because we have started a pull request (#916) to merge this branch into dev.  

We did not achieve our full goal of adding skills to the classes, because we are waiting on the skill-trees team to determine which kinds of skills are possible and supported (in fact, I intend to work with them on that for sprint 3).  However, the machinery to add skills to classes and to test them is present, so adding skills later will be a fairly quick task.
